doc/sphinx/proof-engine/ltac.rst

```diff
@@ -922,25 +1024,87 @@ We can perform pattern matching on goals using the following expression:
-      to backtrack into a right-hand side tactic which has backtracking points
-      left and trigger the selection of a new matching branch or combination of
-      hypotheses when all the backtracking points of the right-hand side have
-      been consumed.
-
       The syntax :n:`match [reverse] goal …` is, in fact, a shorthand for
       :n:`once multimatch [reverse] goal …`.
 
-   .. tacv:: lazymatch goal with {+| {+, @context_hyp} |- @cpattern => @ltac_expr } | _ => @ltac_expr end
+Examples:
+
+.. _match_goal_hyps_ex:
+
+   .. example:: Matching hypotheses
+
+      Hypotheses are matched from the last hypothesis (which is by default the newest
+      hypothesis) to the first until the :n:`apply` succeeds.
+
+      .. coqtop:: reset all
 
-      Using lazymatch instead of match will perform the same pattern matching
-      procedure but will commit to the first matching branch with the first
-      matching combination of hypotheses rather than trying a new matching if
-      the right-hand side fails. If the right-hand side of the selected branch
-      is a tactic with backtracking points, then subsequent failures cause
-      this tactic to backtrack.
+         Goal forall A B : Prop, A -> B -> (A->B).
+         intros.
+         match goal with
+         | H : _ |- _ => idtac "apply " H; apply H
+         end.
+
+.. _match_goal_hyps_rev_ex:
+
+   .. example:: Matching hypotheses with reverse
+
+      Hypotheses are matched from the first hypothesis to the last until the :n:`apply` succeeds.
+
+      .. coqtop:: reset all
+
+         Goal forall A B : Prop, A -> B -> (A->B).
+         intros.
+         match reverse goal with
+         | H : _ |- _ => idtac "apply " H; apply H
+         end.
+
+.. _match_goal_multiple_hyps_ex:
+
+   .. example:: Multiple ways to match hypotheses
+
+      Every possible match for the hypotheses is evaluated until the right-hand
+      side succeeds.  Note that H1 and H2 are never matched to the same hypothesis.
+      Observe that the number of permutations can grow as the factorial
+      of the number of hypotheses and hypothesis patterns.
+
+      .. coqtop:: reset all
+
+         Goal forall A B : Prop, A -> B -> (A->B).
+         intros A B H.
+         match goal with
+         | H1 : _, H2 : _ |- _ => idtac "match " H1 H2; fail
+         | _ => idtac
+         end.
+
+.. _match_vs_lazymatch:
+
+   .. example:: Comparison of match and lazymatch
+
+      .. coqtop:: reset in
+
+         Goal True.
+
+      .. coqtop:: all
+
+         match False with
+         | False => idtac "try"; fail
+         | _ => idtac "success"
+         end.
+
+      **todo: I don't see the "success" line in the output that I see in CoqIDE**
+
+      .. coqtop:: reset in
+
+         Goal True.
+
+      .. coqtop:: all fail
+
+         lazymatch False with
+         | False => idtac "try"; fail
+         | _ => idtac "success"
+         end.
+
+   **Théo, can you give me an example for [ ... ] matching?**
```

Comment:
Which feature do you need an example for ? :)

Reply:
Will clarify in the next update